### DISCUSSION TITLE
Allow shift operators with unsigned right-hand side operands

### DISCUSSION THREAD

**Discussion by philosaurus** [dotnet/csharplang#4431] [Language Ideas]
As the title says: `operator <<` and `>>` are only defined when the right-hand side operand (the number of places to shift by) is an `int`.

This is needlessly restrictive, not to mention it's also _semantically incorrect_, as negative shifts do not follow reversed semantics for negative arguments (a left shift by −1 is _not_ the same as a right shift by 1) but wrap around instead (as unsigned arithmetic does).

I know there is no chance now to fix how shifting by negative values works, but extending the operators to support unsigned operands to avoid having to cast to `int` does not seem like a breaking change.

**Comment by fa1ntl1ght** [comment]
I don't see a point in restricting these operators to have only integers as right-hand operands in general. It's kind of strange to do it, since no other operators are restricted this way. I've heard an opinion that these operators should be restricted to prevent their use for stream operations. But I think it's already clear that it's a bad pattern from the inability to use most of the operators as statements.

**Reply by HaloFour** [reply]
What's a use case to allowing a non-integral type on the RHS that doesn't involve abusing the operator for something other than bitshifting?

**Reply by fa1ntl1ght** [reply]
It still can be bitshifting, but just involving some wrapping constructs.
e.g. when roles and extensions will be added I will be able to do operations on tuples by applying operators to the corresponding values in pairs.

Here's an example use with custom tuple so far:
```cs
public class Program
{
	public static void Main(string[] args)
	{
		MyTuple t1 = new(1, 5);
		MyTuple t2 = new(2, 3);
		Console.WriteLine(t1 + t2);
		Console.WriteLine(t1 - t2);
		Console.WriteLine(t1 * t2);
		Console.WriteLine(t1 / t2);
		Console.WriteLine(t1 & t2);
		Console.WriteLine(t1 | t2);
		Console.WriteLine(t1 ^ t2);
		Console.WriteLine(t1 >> t2);
		Console.WriteLine(t1 << t2);
	}
}

public struct MyTuple
{
	public int A { get; }
	public int B { get; }
	
	public MyTuple(int a, int b)
	{
		A = a;
		B = b;
	}

	public override string ToString() => $"({A}, {B})";

	public static MyTuple operator +(MyTuple t1, MyTuple t2) => new(t1.A + t2.A, t1.B + t2.B);
	public static MyTuple operator -(MyTuple t1, MyTuple t2) => new(t1.A - t2.A, t1.B - t2.B);
	public static MyTuple operator *(MyTuple t1, MyTuple t2) => new(t1.A * t2.A, t1.B * t2.B);
	public static MyTuple operator /(MyTuple t1, MyTuple t2) => new(t1.A / t2.A, t1.B / t2.B);
	public static MyTuple operator &(MyTuple t1, MyTuple t2) => new(t1.A & t2.A, t1.B & t2.B);
	public static MyTuple operator |(MyTuple t1, MyTuple t2) => new(t1.A | t2.A, t1.B | t2.B);
	public static MyTuple operator ^(MyTuple t1, MyTuple t2) => new(t1.A ^ t2.A, t1.B ^ t2.B);
	public static MyTuple operator >>(MyTuple t1, MyTuple t2) => new(t1.A >> t2.A, t1.B >> t2.B);
	public static MyTuple operator <<(MyTuple t1, MyTuple t2) => new(t1.A << t2.A, t1.B << t2.B);
}
```

**Reply by Floris0106** [reply]
I also just had to close [this](https://github.com/Unity-Technologies/Unity.Mathematics/issues/219) feature request, because of this. Vector numerics are the perfect application for not having an int for the rhs, both for Unity and System.Numerics.Vector<T> for example. And besides that, why would it matter if you overloaded the bitshift operator with a different purpose than bitshifting? You can do exactly that with every other operator, so why not with the bitshift operator?

**Reply by CyrusNajmabadi** [reply]
@Floris0106 this is already being done for the next c# release: https://github.com/dotnet/csharplang/blob/main/proposals/relaxing_shift_operator_requirements.md

**Reply by Floris0106** [reply]
Oh, I did not know that, thank you! Which C# version will that be?

**Reply by 333fred** [reply]
11

**Comment by Xyncgas** [comment]
So we can have cout<<"Hello World";
